src/calcium/qqbar.jl

```diff
@@ -0,0 +1,837 @@
+   if !found
+      error("No suitable algebraic number found")
+   end
+   return res
+end
+
+function guess(R::CalciumQQBarField, x::arb, maxdeg::Int, maxbits::Int=0)
+   CC = AcbField(precision(parent(x)))
+   return guess(R, CC(x), maxdeg, maxbits)
+end
+
+###############################################################################
+#
+#   Conversions
+#
+###############################################################################
+
+function (R::ArbField)(a::qqbar)
+   prec = precision(R)
+   z = R()
+   ccall((:qqbar_get_arb, libcalcium), Nothing, (Ref{arb}, Ref{qqbar}, Int), z, a, prec)
+   !isfinite(z) && throw(DomainError(a, "nonreal algebraic number"))
+   return z
+end
+
+function (R::AcbField)(a::qqbar)
+   prec = precision(R)
+   z = R()
+   ccall((:qqbar_get_acb, libcalcium), Nothing, (Ref{acb}, Ref{qqbar}, Int), z, a, prec)
+   return z
+end
+
+# todo: provide qqbar_get_fmpq, qqbar_get_fmpz in C
+function fmpq(a::qqbar)
+   !isrational(a) && throw(DomainError(a), "nonrational algebraic number")
+   p = fmpz()
+   q = fmpz()
+   ccall((:fmpz_poly_get_coeff_fmpz, libflint), Nothing, (Ref{fmpz}, Ref{qqbar}, Int), p, a, 0)
+   ccall((:fmpz_poly_get_coeff_fmpz, libflint), Nothing, (Ref{fmpz}, Ref{qqbar}, Int), q, a, 1)
+   ccall((:fmpz_neg, libflint), Nothing, (Ref{fmpz}, Ref{fmpz}), p, p)
+   return p // q
+end
+
+function fmpz(a::qqbar)
+   !isinteger(a) && throw(DomainError(a), "noninteger algebraic number")
+   z = fmpz()
+   ccall((:fmpz_poly_get_coeff_fmpz, libflint), Nothing, (Ref{fmpz}, Ref{qqbar}, Int), z, a, 0)
+   ccall((:fmpz_neg, libflint), Nothing, (Ref{fmpz}, Ref{fmpz}), z, z)
+   return z
+end
+
+###############################################################################
+#
+#   Unsafe functions
+#
+###############################################################################
+
+function zero!(z::qqbar)
+   ccall((:qqbar_zero, libcalcium), Nothing, (Ref{qqbar},), z)
+   return z
+end
+
+function mul!(z::qqbar, x::qqbar, y::qqbar)
+   ccall((:qqbar_mul, libcalcium), Nothing,
+                (Ref{qqbar}, Ref{qqbar}, Ref{qqbar}), z, x, y)
+   return z
+end
+
+function addeq!(z::qqbar, x::qqbar)
+   ccall((:qqbar_add, libcalcium), Nothing,
+                (Ref{qqbar}, Ref{qqbar}, Ref{qqbar}), z, z, x)
+   return z
+end
+
+function add!(z::qqbar, x::qqbar, y::qqbar)
+   ccall((:qqbar_add, libcalcium), Nothing,
+                (Ref{qqbar}, Ref{qqbar}, Ref{qqbar}), z, x, y)
+   return z
+end
+
+###############################################################################
+#
+#   Parent object call overloads
+#
+###############################################################################
+
+(a::CalciumQQBarField)() = qqbar()
+
+(a::CalciumQQBarField)(b::Int) = qqbar(b)
+
+(a::CalciumQQBarField)(b::fmpz) = qqbar(b)
+
+(a::CalciumQQBarField)(b::fmpq) = qqbar(b)
+
+(a::CalciumQQBarField)(b::qqbar) = b
```

Comment:
Any possibility to add a single line test for this under constructors or similar. eg. just test it has the right type.